types/tx/middleware.go

```diff
@@ -0,0 +1,14 @@
+package tx
+
+import (
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type TxHandler interface {
+	CheckTx(ctx sdk.Context, tx sdk.Tx, req abci.RequestCheckTx) (abci.ResponseCheckTx, error)
```

Comment:
Maybe let's use `context.Context`? There have been discussions about deprecating `sdk.Context` and I think that should generally be the direction. 

And maybe we also use the concrete Tx type and not the sdk.Tx interface. I see no need to support StdTx or any other tx type for that matter other than for signing.

Reply:
OK for `context.Context`.

For sdk.Tx versus proto Tx, baseapp's txDecoder returns a `sdk.Tx`, which is a txWrapper. How do I get the proto Tx from there?

I bumped into this issue at least once before, and the (unclean) solution that time was to cast to a `ProtoTxProvider` interface, which `txwrapper` implements. Currently, `ProtoTxProvider` is a private test interface. Should we make it public so that we can do:
```go
sdkTx := baseapp.txDecoder(txBytes)
protoTx := sdkTx.(ProtoTxProvider).GetProtoTx()
```

[see current ProtoTxProvider](https://github.com/cosmos/cosmos-sdk/blob/cd221680c0509a39b539d3dd7e3e45d700aa4b54/x/auth/tx/service_test.go#L606-L612)

Reply:
I think we might want to make the BaseApp tx decoder less generic... In general though, what do you think about this direction? I think it's probably something we should bring up in the next architecture call to be sure...

My thinking is that middleware will be simpler and more predictable if it's based on the concrete tx type, and also this generic tx interface stuff is a source of pain it seems. On the other hand middleware is more composable between different tx types if we do the current tx interface casting

And regarding context, it seems like the way to allow more flexibility at the infrastructure layer (with different types of storage) is to not have `sdk.Context` as a bag of variables but rather just use `context.Context`

Reply:
👍 for using the concrete proto Tx type here. But for this PR, let's keep sdk.Tx for now, and raise the issue of removing sdk.Tx in our next architecture call.

As for deprecating sdk.Context, overall okay, but i'm missing a bit of context. Would love to see a discussion about this (around Froydi's idea).